src/cart/crt_hg.h

```diff
@@ -172,13 +172,19 @@ crt_hgret_2_der(int hg_ret)
 	case HG_INVALID_ARG:
 		return -DER_INVAL;
 	case HG_MSGSIZE:
+	case HG_OVERFLOW:
 		return -DER_OVERFLOW;
 	case HG_NOMEM:
 		return -DER_NOMEM;
 	case HG_CANCELED:
 		return -DER_CANCELED;
 	case HG_BUSY:
 		return -DER_BUSY;
+	case HG_PROTOCOL_ERROR:
+		return -DER_PROTO;
```

Comment:
These two aren't the same.  I think maybe HG_PROTOCOL_ERROR should match DER_UNREG and DER_PROTO doesn't have a direct equivalent but you would need to check the headers to confirm.

Reply:
> These two aren't the same. I think maybe HG_PROTOCOL_ERROR should match DER_UNREG and DER_PROTO doesn't have a direct equivalent but you would need to check the headers to confirm.

the ticket's case the HG_Bulk_create() failed as HG_PROTOCOL_ERROR, the most match things in my mind is DER_PROTO.
DER_UNREG mostly used for RPC unregistered.

Reply:
Maybe HG_PROTOCOL_ERROR should match DER_PROTO then as they're both unexpected input or problems with packing/unpacking.  Either this would mean a bug on the sender/receiver or the remote end shouldn't be trusted so I assume this is not recoverable?

HG_PROTONOSUPPORT should match DER_UNREG as they both mean RPC unregistered.

@soumagne would be best to review this.

Reply:
`HG_PROTOCOL_ERROR` is really a catch-all generic type of error. From your description I felt originally that `DER_PROTO` better matches `HG_PROTONOSUPPORT`. I'd leave `HG_PROTOCOL_ERROR` return `DER_HG` I think or something generic that means something unrecoverable happened.

Reply:
@soumagne  the ticket reported failure is the HG_Bulk_create() failed as 12 which is HG_PROTOCOL_ERROR, so need to make it to be un-retryable err. Then this patch convert HG_Bulk_create to DER_PROTO to satisfy that requirement. For other cases HG return HG_PROTOCOL_ERROR, looks all belong to un-retryable case as well.
The generic err code DER_HG is still retryable network err code used for other purpose.
So if you leave HG_PROTOCOL_ERROR return DER_HG, the original dead loop cannot resolve.
The current method is by crt_hgret_2_der() it convert the un-retryable err code to specific err code, and other cases still use DER_HG that is retryable. some details probably can be refined later if hit some special case in test, but that probably is simplest and safe way to resolve the problem now.

I think it looks natural to match HG_PROTOCOL_ERROR to DER_PROTO, if that is not good probably better to change mercury code to not return HG_PROTOCOL_ERROR for that case.

HG_PROTONOSUPPORT never hit in daos tests before, I am not very sure for that case, please feel free to submit other patch for that later if needed.

Reply:
Ok fair enough. I was not thinking that DER_HG was considered as retryable. In that case yes it's probably better to make it match DER_PROTO for now. Although I think it would be better in the long term to keep them separate since PROTONOSUPPORT means that eventually you could fallback to another method that is supported.